### PR TITLE
REALMC-8260: Add template app support for realm-cli app create

### PR DESCRIPTION
**Pull request by UnicodeSnowman**

### CODE REVIEW

File: internal/commands/app/create.go [hunk 1]
```diff
@@ -43,6 +49,7 @@ func (cmd *CommandCreate) Flags(fs *pflag.FlagSet) {
 	fs.VarP(&cmd.inputs.Environment, flagEnvironment, flagEnvironmentShort, flagEnvironmentUsage)
 	fs.StringVar(&cmd.inputs.Cluster, flagCluster, "", flagClusterUsage)
 	fs.StringVar(&cmd.inputs.DataLake, flagDataLake, "", flagDataLakeUsage)
+	fs.StringVar(&cmd.inputs.Template, flagTemplate, "", flagTemplateUsage)
```
Comment (UnicodeSnowman) [inline internal/commands/app/create.go:52]: went with this single flag instead of the double `--template` and `--use-template` flags described in the ticket (see `resolveTemplateID` for usage) because this seemed simpler, but willing to discuss if there was a particular reason for that.

**Comment by UnicodeSnowman** [comment]
reminder to myself: I should merge this into a feature branch for now until we actually want to release this

**Comment by makesitgo** [comment]
in regards to feature branch and such, this conversation may be related: https://github.com/10gen/realm-cli/pull/156#discussion_r646622096

just want to call out we have options here (other than long-lived branches) for allowing this code to live before it wants to be released

File: internal/commands/app/create_inputs.go [hunk 2]
```diff
@@ -131,6 +135,65 @@ func (i *createInputs) resolveLocalPath(ui terminal.UI, wd string) (string, erro
...
+	// available and user is not specifying a template
+	if templateID == "" && len(templates) == 0 {
+		return "", nil
+	}
+
+	if len(templates) == 0 {
+		return "", fmt.Errorf("unable to find template '%s'", templateID)
+	}
+
+	var resolvedTemplateID string
```
Comment (makesitgo) [inline internal/commands/app/create_inputs.go:158]: could reduce a little indenting and likely avoid this `resolvedTemplateID` var all together:

```go
if templateID != "" {
  for _, template := range templates {
    if template.ID == i.Template {
      return template.ID, nil
    }
  }
  return "", fmt.Errorf("template '%s' not found", templateID)
}

options := ...
templateIDs := ...

// ui.AskOne...
```

File: internal/commands/app/create_inputs.go [hunk 3]
```diff
@@ -131,6 +135,65 @@ func (i *createInputs) resolveLocalPath(ui terminal.UI, wd string) (string, erro
 	return fullPath, nil
 }
 
+func (i *createInputs) resolveTemplateID(ui terminal.UI, client realm.Client, templateID string) (string, error) {
```
Comment (makesitgo) [inline internal/commands/app/create_inputs.go:138]: worth adding any unit tests here?  just to verify the various errors/prompts that can occur here

File: internal/commands/app/create_test.go [hunk 4]
```diff
@@ -393,6 +402,96 @@ func TestAppCreateHandler(t *testing.T) {
...
+		assert.Nil(t, err)
+		assert.Equal(t, len(backendFileInfo), 10)
+
+		frontendFileInfo, err := ioutil.ReadDir(filepath.Join(profile.WorkingDirectory, cmd.inputs.Name, frontendPath))
+		assert.Nil(t, err)
+		assert.Equal(t, len(frontendFileInfo), 1)
+		assert.Equal(t, frontendFileInfo[0].Name(), "react-native")
+
+		dirLength := len(appLocal.RootDir)
+		fmt.Printf("%#v\n", appLocal.RootDir)
```
Comment (makesitgo) [inline internal/commands/app/create_test.go:480]: not needed?

File: internal/commands/app/create_test.go [hunk 5]
```diff
@@ -702,7 +859,9 @@ func TestAppCreateHandler(t *testing.T) {
 				DataLake: tc.dataLake,
 			}}
 
-			assert.Equal(t, tc.expectedErr, cmd.Handler(profile, nil, tc.clients))
+			_, ui := mock.NewUI()
```
Comment (makesitgo) [inline internal/commands/app/create_test.go:862]: what's the `ui` used for here?

File: internal/commands/app/new_app_inputs.go [hunk 6]
```diff
@@ -57,6 +57,7 @@ type newAppInputs struct {
 	DeploymentModel realm.DeploymentModel
 	Location        realm.Location
 	Environment     realm.Environment
+	Template        string
```
Comment (makesitgo) [inline internal/commands/app/new_app_inputs.go:60]: would we want `resolveTemplateID` defined on `newAppInputs` instead of `createInputs`?  also, do we want template behavior for `app init`?

**Review by makesitgo — COMMENTED** [review]
just some various thoughts/musings/questions here

File: internal/cloud/realm/templates.go [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+	res, resErr := c.do(
+		http.MethodGet,
+		templatesPath,
+		api.RequestOptions{},
+	)
+	if resErr != nil {
+		return nil, resErr
+	}
+	if res.StatusCode != http.StatusOK {
+		return nil, api.ErrUnexpectedStatusCode{"fetch templates", res.StatusCode}
```
[the remark this answers is not in the document]
Comment (UnicodeSnowman) [reply internal/cloud/realm/templates.go:35]: yeah, happy to be consistent here

File: internal/cloud/realm/app.go [hunk 8]
```diff
@@ -21,6 +21,7 @@ type AppMeta struct {
 	Location        Location        `json:"location,omitempty"`
 	DeploymentModel DeploymentModel `json:"deployment_model,omitempty"`
 	Environment     Environment     `json:"environment,omitempty"`
+	Template        string          `json:"template_id,omitempty"`
```
[the remark this answers is not in the document]
Comment (UnicodeSnowman) [reply internal/cloud/realm/app.go:24]: we unfortunately don't have this enabled in a "real" environment yet, but that's certainly something we should do at some point. @andrewzlchen is there a general "E2E Testing" ticket or anything of that sort? Although, we spin up a Realm server for CLI integration tests right? we could just enable templates in the config if that's the case. I'll investigate.

File: internal/commands/app/create.go [hunk 9]
```diff
@@ -13,6 +14,11 @@ import (
 	"github.com/spf13/pflag"
 )
 
+var (
```
[the remark this answers is not in the document]
Comment (UnicodeSnowman) [reply internal/commands/app/create.go:17]: my bad

File: internal/commands/app/create.go [hunk 10]
```diff
@@ -156,15 +190,32 @@ func (cmd *CommandCreate) Handler(profile *user.Profile, ui terminal.UI, clients
...
-		appLocal, err = local.LoadApp(dir)
+
+		appLocal, err = local.LoadApp(backendDir)
 		if err != nil {
 			return err
 		}
 	}
 
+	if templateID != "" {
+		zipPkg, err := clients.Realm.ClientTemplate(
```
[the remark this answers is not in the document]
Comment (UnicodeSnowman) [reply internal/commands/app/create.go:205]: yeah good point, I'll look into the spinner, this could definitely take a bit if the template is grande

File: internal/commands/app/create_inputs.go [hunk 11]
```diff
@@ -131,6 +135,65 @@ func (i *createInputs) resolveLocalPath(ui terminal.UI, wd string) (string, erro
...
+	if templateID == "" && ui.AutoConfirm() {
```
[the remark this answers is not in the document]
Comment (UnicodeSnowman) [reply internal/commands/app/create_inputs.go:139]: yeah, but then how would you handle the case where a user is using `-y` (perhaps already, in a CI flow or something) without breaking things for them when we introduce this feature? if the requirement is that we always prompt a user for a template option _unless_ they've specified a `--template` param, we want to be able to skip the prompt/required param for those users using `-y`. otherwise, we either force them to provide a `--template none` (which doesn't feel right).

File: internal/commands/app/create_inputs.go [hunk 12]
```diff
@@ -131,6 +135,65 @@ func (i *createInputs) resolveLocalPath(ui terminal.UI, wd string) (string, erro
...
+
+	if len(templates) == 0 {
+		return "", fmt.Errorf("unable to find template '%s'", templateID)
+	}
+
+	var resolvedTemplateID string
+	if templateID == "" {
+		options := make([]string, 0, len(templates)+1)
+		templateIDs := make([]string, 0, len(templates)+1)
+		options = append(options, "[No Template]: Do Not Use A Template")
```
[the remark this answers is not in the document]
Comment (UnicodeSnowman) [reply internal/commands/app/create_inputs.go:162]: yeah, but "every time" here is really only when a user is creating a new app, which seems completely fine here. otherwise, if a user is using `-y`, we skip all of this, unless said `-y` user has specified a `--template` param.

the other option is to **not** automatically/interactively prompt users to select a template, but I feel like that makes the template apps feature less discoverable for CLI users. @andrewzlchen might have more context on that requirement though

File: internal/commands/app/create_inputs.go (hunk 3 above)
Comment (UnicodeSnowman) [reply internal/commands/app/create_inputs.go:138]: unit tests for what specifically? We definitely don't want to directly unit test `resolveTemplateID` since it's a private method, but I can likely update my existing test cases to try to hit some of these additional branches.

File: internal/commands/app/create_test.go (hunk 5 above)
Comment (UnicodeSnowman) [reply internal/commands/app/create_test.go:862]: it's passed to the next line

File: internal/commands/app/new_app_inputs.go (hunk 6 above)
Comment (UnicodeSnowman) [reply internal/commands/app/new_app_inputs.go:60]: > would we want resolveTemplateID defined on newAppInputs instead of createInputs?

yeah, actually it probably makes more sense there, good call 👍 

> also, do we want template behavior for app init?

truuuue, I forgot about that command, also a good call

File: internal/commands/app/create_test.go (hunk 5 above)
Comment (makesitgo) [reply internal/commands/app/create_test.go:862]: ah the `AutoConfirm` check, roger that!

File: internal/commands/app/create_inputs.go (hunk 3 above)
Comment (makesitgo) [reply internal/commands/app/create_inputs.go:138]: wait what's this "dont unit test private methods" idea.... have I been doing code wrong??

I'm thinking tests, somewhere, that at least prove the errors we return (any `Templates` error, the unable to find template error, the template not found error) as well a test that verifies our behavior when `--template` is not set but `--yes` is, behavior when `--template` is not set and no templates are found, and that the prompt eventually shows up and lets us pick an option

File: internal/commands/app/new_app_inputs.go (hunk 6 above)
Comment (andrewzlchen) [reply internal/commands/app/new_app_inputs.go:60]: hmm i think `app init` may probably have been missed during scoping but it certainly does make sense to support templates for app init, so I guess this `Template` field can stay here

Since it doesn't seem strictly necessary, i think we can skip the init command template implementation for the time being and then come back to it once we get some breathing room when AB test goes out

File: internal/commands/app/create_inputs.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (andrewzlchen) [reply internal/commands/app/create_inputs.go:162]: Hmm i think we should prompt users on the template options on every `realm-cli app create` invokation

Reason for that is that the objective of this project to help users who are starting out with realm to have a gentle intro experience. 

Hiding the templates by default kinda goes against that project objective.

This was also kinda why i wanted to use the `--use-template` flag because we can skip fetching the templates in the event the user says no which may be easier on the server

The bottom line is that we should definitely prompt for whether the user wants to create their app with a template - whether it's with the --use-template flag or the just the --template

**Comment by andrewzlchen** [comment]
I dont have much to say about the intricacies of the command implementation itself, but left some comments about my thoughts about the functionality

File: internal/commands/app/create_inputs.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (andrewzlchen) [reply internal/commands/app/create_inputs.go:162]: Not sure if I'm misinterpreting your intent, but my understanding of the CLI was that it would prompt in the event that the user does not pass in a required flag. If this is true then we would need to prompt if the user does not supply the template parameter out of necessity.

Also, I don't really want to introduce new concepts here by introducing a global flag tbh

Maybe the happy middle ground here is to have the `--use-template` (or a `--no-template` since --use-template is only needed if we didn't want to use a template)

File: internal/commands/app/create_inputs.go (hunk 3 above)
Comment (UnicodeSnowman) [reply internal/commands/app/create_inputs.go:138]: I'm happy to test that behavior through the execution of the main commands itself, but testing private methods is pretty much always a bad idea unless there's a very specific reason for doing so. If someone changes the name of a private function that you're testing or makes some other internal refactoring change that doesn't actually alter the way the command behaves from the user's perspective, it's much more likely that that test will need to be updated to reflect the "new" version of the private behavior. It's a recipe for brittle tests.

File: internal/cloud/realm/client.go [hunk 13]
```diff
@@ -70,6 +70,9 @@ type Client interface {
 
 	SchemaModels(groupID, appID, language string) ([]SchemaModel, error)
 
+	Templates() ([]Template, error)
```
[the remark this answers is not in the document]
Comment (UnicodeSnowman) [reply internal/cloud/realm/client.go:73]: Wait, what? Why a query param? The templates endpoint returns a list of of `{ id, name }` for the available template options. If you're just querying for a single template that should be `/templates/:template_id`.

[the remark this answers is not in the document]
Comment (andrewzlchen) [reply internal/cloud/realm/client.go:73]: Oh no, the function should still return a list of Templates.

The update to the endpoint is so we can return the compatible templates for the template id that we pass in

[the remark this answers is not in the document]
Comment (andrewzlchen) [reply internal/cloud/realm/client.go:73]: https://jira.mongodb.org/browse/REALMC-9191

File: internal/commands/app/create_inputs.go (hunk 3 above)
Comment (makesitgo) [reply internal/commands/app/create_inputs.go:138]: This whole codebase is brittle-y tested then... :bushes:

The idea I had in mind while writing tests for this CLI was: separate out inputs handling and command functionality.  I would still argue there's great value in unit testing our private, inputs-related functions: mostly around the various ways we intend to resolve said inputs.  in effect, it allowed us to, in detail, test the interactivity of CLI prompts and collecting inputs separately from the actual `Handler` implementation.  You could definitely just write these tests as `Handler` tests, but in my opinion that result is a bit overkill.

either way, it would be good to have something here that proves out the various errors we're returning, and the instances we actual prompt for more info

File: internal/cloud/realm/app.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (makesitgo) [reply internal/cloud/realm/app.go:24]: yeah you have the `test_cli_config` to play around with.  as long as the endpoints live in `master` you should be able to get what you need spun up for the server.  you don't need to go crazy here, but its been useful for proving out we have the right endpoints, requests, and responses

File: internal/commands/app/create_inputs.go (hunk 3 above)
Comment (UnicodeSnowman) [reply internal/commands/app/create_inputs.go:138]: yeah, I'll always prefer to see those tests at the `Handler` level since there doesn't seem to be much (anything?) you _can't_ test from that level, and it avoids the issue of coupling your tests to implementation details vs. end user behavior. I don't enjoy "fixing" tests for things that aren't actually broken.

File: internal/cloud/realm/client.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (UnicodeSnowman) [reply internal/cloud/realm/client.go:73]: so the `?template_id=ios.swift.todo` refers to a backend template id? or frontend template id? I'm a bit confused by the idea of a "compatible template" in this context. if you already have the id of a template, why would there be multiple results in a call for that template id? in your example, it seems like there would somehow be multiple `ios.swift.todo` templates?

[the remark this answers is not in the document]
Comment (UnicodeSnowman) [reply internal/cloud/realm/client.go:73]: I feel like that type of endpoint belongs under the `/groups/%s/apps/%s/templates` namespace rather than the `/templates` namespace... the latter is really just a list of all available templates, whereas the former is already tied to the app in question... you wouldn't even need to pass a template id if you put this under `/groups/%s/apps/%s/templates`, since you'll already know the template id the app was created with in the handler.

[the remark this answers is not in the document]
Comment (andrewzlchen) [reply internal/cloud/realm/client.go:73]: Yeah, sorry for any confusion - lots of specific lingo getting thrown around here.

So when I refer to a template, I mean the `templateapps.TemplateApp` in the backend or basically a coupling of a backend and an optional frontend.

There is no concept of "frontend template" or "backend template". A template id refers to the ID of a unique coupling of the backend and client - if one exists.

The reason why we want to provide the template_id to the `/templates` endpoint is so we can see which templates among those found in the templates cache share the same backend as the one passed in. We don't need to do this in an app-level endpoint handler because determining which templates have the same backends can be done without knowing anything about the app that was created from the template.

As a concrete example, the `ios.swift.todo` (found [here](https://github.com/mongodb-university/realm-template-apps/blob/02b5d4a9610cdf7d8dc7b5532c156af78c9dba44/manifest.json#L2-L11)) and the `android.kotlin.todo` templates (found [here](https://github.com/mongodb-university/realm-template-apps/blob/02b5d4a9610cdf7d8dc7b5532c156af78c9dba44/manifest.json#L12-L21)) both have the same backend (the same repository: "github.com/mongodb-university/realm-template-apps" and the same backend path: "todo-sync").

This means that the two templates have the same backend but have different frontends.

Since this is the case, it is totally valid to have a Realm app created with, say the `ios.swift.todo` template to support exporting the frontend for the  `android.kotlin.todo` template since the 2 frontends were built with the same realm app as their serverless backends.

This is the use-case that we're trying to support by adding the query parameters to the `/templates` endpoint. I still think that we should add the query param instead of the new endpoint handler because we are still listing templates, but we're filtering out the ones that do not have the same backend as the template that is passed in

sorry for wall of text btw

File: internal/commands/app/create_inputs.go (hunk 3 above)
Comment (makesitgo) [reply internal/commands/app/create_inputs.go]: a lot of the input resolution in commands will happen before `Handler` is ever called, not really the case here with `resolveTemplateID`.  but lets at least have tests that flush out the various behaviors of `resolveTemplateID`, either with a test directly on that function (which I really still find to be fine here), or as variants of a tests on the `Handler` method.  basically the error cases and the interactive prompt case (and for the interactive test, you'll want to use a different form of the "mock ui", of which there are plenty of examples of)

File: internal/cloud/realm/client.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (UnicodeSnowman) [reply internal/cloud/realm/client.go:73]: chatted on Slack about this, not going to do anything here yet for this.

**Comment by UnicodeSnowman** [comment]
@andrewzlchen @makesitgo ready for another look.

I'll likely comment out the code that automatically prompts users for a template if not provided before merging this (so that we can release the backend code for template starter apps to prod before the UI goes out) and link a REALMC ticket to uncomment after GA

**Review by andrewzlchen — APPROVED** [review]
Just had small nits but this LGTM overall!

File: internal/commands/app/create.go [hunk 14]
```diff
@@ -101,13 +114,29 @@ func (cmd *CommandCreate) Handler(profile *user.Profile, ui terminal.UI, clients
...
+	// backend code there alongside a sibling directory containing the frontend
+	// code. Otherwise, all code is exported in rootDir
+	backendDir := rootDir
+	if templateID != "" {
+		backendDir = path.Join(rootDir, backendPath)
+	}
+
 	if cmd.inputs.DryRun {
 		logs := make([]terminal.Log, 0, 4)
+		var templateIDText string
```
[the remark this answers is not in the document]
Comment (UnicodeSnowman) [reply internal/commands/app/create.go:127]: Whoops yep, that variable was left over from a previous implementation of this, good catch

File: internal/commands/app/new_app_inputs.go [hunk 15]
```diff
@@ -62,3 +65,56 @@ func (i *newAppInputs) resolveRemoteApp(ui terminal.UI, rc realm.Client) (realm.
...
+		return "", err
+	}
+
+	// do not disrupt application creation flow if templates are not
+	// available and user is not specifying a template
+	if i.Template == "" && len(templates) == 0 {
+		return "", nil
+	}
+
+	if len(templates) == 0 {
```
[the remark this answers is not in the document]
Comment (UnicodeSnowman) [reply internal/commands/app/new_app_inputs.go:85]: > Without the --use-template flag and when the /templates endpoint doesn't return anything, the user cannot create any apps because it'll stop here since we currently always prompt the user for the template id

I reeeeally want to avoid ever implementing something like `--use-template`, because it's such a single-use escape hatch feature... the only time you would ever use it would be to set `--use-template false`, which feels extremely clunky to me.

here, we only error out if we failed to fetch templates _and_ a user has specified a template param. if they don't specify a template, we return `""` above on line 82, allowing the user to continue as usual and create non-template apps. if we don't have any templates, there's really no way forward for creating a templated app, but they can still create a regular "blank" app.

**Comment by UnicodeSnowman** [comment]
tests currently broken due to config issues, https://github.com/10gen/baas/pull/4342 should fix this

File: internal/commands/app/create.go [hunk 16]
```diff
@@ -43,8 +51,8 @@ func (cmd *CommandCreate) Flags(fs *pflag.FlagSet) {
 	fs.VarP(&cmd.inputs.Environment, flagEnvironment, flagEnvironmentShort, flagEnvironmentUsage)
 	fs.StringVar(&cmd.inputs.Cluster, flagCluster, "", flagClusterUsage)
 	fs.StringVar(&cmd.inputs.DataLake, flagDataLake, "", flagDataLakeUsage)
+	fs.StringVar(&cmd.inputs.Template, flagTemplate, "", flagTemplateUsage)
 	fs.BoolVarP(&cmd.inputs.DryRun, flagDryRun, flagDryRunShort, false, flagDryRunUsage)
-
```
Comment (makesitgo) [inline internal/commands/app/create.go:47]: fwiw, I had this newline to try and _separate_ out public v private/hidden flags.  no action needed here, i'm fine with either approach.  there's likely going to be a mix of both styles in the codebase eventually

File: internal/commands/app/new_app_inputs.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (andrewzlchen) [reply internal/commands/app/new_app_inputs.go:85]: ah okay yeah i see now. I'm good with this then - feel free to ignore!

File: internal/commands/app/create.go [hunk 17]
```diff
@@ -122,7 +151,12 @@ func (cmd *CommandCreate) Handler(profile *user.Profile, ui terminal.UI, clients
 	appRealm, err := clients.Realm.CreateApp(
 		groupID,
 		cmd.inputs.Name,
-		realm.AppMeta{cmd.inputs.Location, cmd.inputs.DeploymentModel, cmd.inputs.Environment},
+		realm.AppMeta{
+			cmd.inputs.Location,
+			cmd.inputs.DeploymentModel,
+			cmd.inputs.Environment,
+			templateID,
```
Comment (makesitgo) [inline internal/commands/app/create.go:158]: I can't comment on the line exactly [L#146](https://github.com/10gen/realm-cli/pull/163/files#diff-56dff07c87480c9d173d72d506ffac3e9e906a3ff89db6be5d819a1143655c63R146), but the `cmd.display` function relies on `inputs` having the values desired to be played back to the user (if run with `--dry-run`)

for that reason, I think it may be desirable to mutate the value of `cmd.inputs.Template` rather than return `templateID` from the `inputs.resolveTemplateID` function.  it's actually a very minor use-case/unlikely scenario here, and the command printed out here would still "work" (just assuming the user resolves the template id the same way if through prompts).

File: internal/commands/app/create_test.go [hunk 18]
```diff
@@ -310,6 +316,82 @@ func TestAppCreateHandler(t *testing.T) {
...
+			DeploymentModel: realm.DeploymentModelGlobal,
+			ConfigVersion:   realm.DefaultAppConfigVersion,
+		}}}
+
+		assert.Nil(t, cmd.Handler(profile, ui, cli.Clients{Realm: rc, Atlas: ac}))
+
+		console.Tty().Close() // flush the writers
+		<-doneCh              // wait for procedure to complete
+
+		_, err = local.LoadApp(filepath.Join(profile.WorkingDirectory, cmd.inputs.Name))
```
Comment (makesitgo) [inline internal/commands/app/create_test.go:391]: is this meant to prove the app was written?  if so, you'll want an assert on `appLocal.RootDir` to make sure its not blank as well here
